bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,206 @@
+from collections.abc import Iterable
+
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from pydantic.utils import deep_update
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.start_stop_node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._sr_nodes = {}
+        self._sr_cache = {}
+
+    def start(self, doc):
+        self.start_stop_node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        doc = dict(doc)
+        doc.pop("start", None)
+        metadata = dict(self.start_stop_node.metadata) | {"stop": doc}
+        self.start_stop_node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        desc_name = doc["name"]
+        metadata = dict(doc)
+
+        # Extract the variable fileds of the descriptor that can change during the run
+        uid = metadata.pop("uid")
+        conf_dict = {uid: metadata.pop("configuration", {})}
+        time_dict = {uid: metadata.pop("time", None)}
```

Comment:
All EventDescriptor documents have always had `time`, so it's same to assume this will be present and that the values in `time_dict` will never be `None`.

```suggestion
        time_dict = {uid: metadata.pop("time")}
```

Reply:
Great, that's what I thought. Just added `None` to err on the safe side.